### PR TITLE
CodeCamp #1492: Support Boxinst

### PR DESCRIPTION
**Pull request by JosonChan1998**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support Boxinst in mmdet3.x, the reproduce precision is aligned with paper:
MS_R_50_1x| box AP | mask AP |
--- |:---:|:---:|
origin   | 39.4 | 30.7 | 
ours(90k)| 39.4 | 30.8 |

## Modification

configs/boxinst
mmdet/models/data_preprocessors
mmdet/models/detectors
mmdet/models/dense_heads

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/dense_heads/boxinst_head.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                head.
+
+        Returns:
+            dict[str, Tensor]: A dictionary of loss components.
+        """
+        assert positive_infos is not None, \
+            'positive_infos should not be None in `BoxInstMaskHead`'
+        losses = dict()
+
+        self._iter += 1
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/boxinst_head.py:105]: we can get the iteration number from message_hub and do not need to calculate it in model

File: mmdet/models/utils/misc.py [hunk 2]
```diff
@@ -625,3 +625,28 @@ def aligned_bilinear(tensor: Tensor, factor: int) -> Tensor:
         tensor, pad=(factor // 2, 0, factor // 2, 0), mode='replicate')
 
     return tensor[:, :, :oh - 1, :ow - 1]
+
+
+def unfold_wo_center(x, kernel_size, dilation):
+    """unfold_wo_center, used in original implement in BoxInst:
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:631]: unfoled_wo_center -> Unfold without xx center?

Comment (JosonChan1998) [reply mmdet/models/utils/misc.py:631]: Yes. without the pixel center.

File: mmdet/models/dense_heads/boxinst_head.py (hunk 1 above)
Comment (JosonChan1998) [reply mmdet/models/dense_heads/boxinst_head.py:105]: Could you please give me an example?

File: mmdet/models/dense_heads/boxinst_head.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            loss_mask_project += loss_project
+            loss_mask_pairwise += loss_pairwise
+
+        if total_pos == 0:
+            total_pos += 1  # avoid nan
+        if avg_fatcor == 0:
+            avg_fatcor += 1  # avoid nan
+        loss_mask_project = loss_mask_project / total_pos
+        loss_mask_pairwise = loss_mask_pairwise / avg_fatcor
+        warmup_factor = min(self._iter.item() / float(self.warmup_iters), 1.0)
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/boxinst_head.py:148]: You can get the training iteration with the following code:
```python
from mmengine import MessageHub
message_hub = MessageHub.get_current_instance()
iter = message_hub.get_info('iter')
```

File: mmdet/models/data_preprocessors/data_preprocessor.py [hunk 4]
```diff
@@ -645,3 +648,126 @@ def get_padded_tensor(self, tensor: Tensor, pad_value: int) -> Tensor:
...
+
+    Args:
+        mask_stride (int): The mask output stride in boxinst. Defaults to 4.
+        pairwise_size (int): The size of neighborhood for each pixel.
+            Defaults to 3.
+        pairwise_dilation (int): The dilation of neighborhood for each pixel.
+            Defaults to 2.
+        pairwise_color_thresh (float): The thresh of image color similarity.
+            Defaults to 0.3.
+        bottom_pixels_removed (int): The length of removed pixels in bottom.
```
Comment (RangiLyu) [inline mmdet/models/data_preprocessors/data_preprocessor.py:670]: Explain in the docstring that it is caused by the annotation error in the coco dataset.

File: mmdet/models/dense_heads/boxinst_head.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+            - pos_pairwise_masks (Tensor): Positive pairwise masks with
+              shape: (num_pos, num_neighborhood, mask_h, mask_w).
+            - num_pos (int): Positive numbers.
+        """
+        gt_bboxes = gt_instances.bboxes
+        device = gt_bboxes.device
+        gt_masks = gt_instances.masks.to_tensor(
+            dtype=torch.bool, device=device).float()
+        pairwise_masks = gt_instances.pairwise_masks
+        pairwise_masks = pairwise_masks.to(device=device)
```
Comment (RangiLyu) [inline mmdet/models/dense_heads/boxinst_head.py:195]: Add comments: these pseudo masks are generated from BoxInstDataPreprocessor.

**Review by RangiLyu — COMMENTED** [review]
The code looks good to me now. You can add some unit tests.

File: mmdet/models/data_preprocessors/data_preprocessor.py [hunk 6]
```diff
@@ -12,11 +12,14 @@
 from mmengine.model import BaseDataPreprocessor, ImgDataPreprocessor
 from mmengine.structures import PixelData
 from mmengine.utils import is_list_of
+from skimage import color
```
Comment (RangiLyu) [inline mmdet/models/data_preprocessors/data_preprocessor.py:15]: skimage is not the default requirement package of mmdet. You should import it like this:

```
try:
    import skimage
except ImportError:
    skimage = None
```

and check it in the __init__ of BoxInstDataPreprocessor like this:
```
    def __init__(self, ...):
        if skimage is None:
            raise RuntimeError(
                'skimage is not installed, please install it by: ...'
```

Comment (JosonChan1998) [reply mmdet/models/data_preprocessors/data_preprocessor.py:15]: Maybe we can use cv2?

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**